test/jdk/java/awt/color/NonICCFilterTest.java

```diff
@@ -0,0 +1,157 @@
+        ComponentColorModel cm = new ComponentColorModel(cs, false, false,
+                Transparency.OPAQUE, DataBuffer.TYPE_BYTE);
+        WritableRaster raster = cm.createCompatibleWritableRaster(WIDTH, HEIGHT);
+        BufferedImage img = new BufferedImage(cm, raster, false, null);
+
+        Graphics2D g = img.createGraphics();
+        GradientPaint gp = new GradientPaint(0, 0, Color.GREEN,
+                raster.getWidth(), raster.getHeight(), Color.BLUE);
+        g.setPaint(gp);
+        g.fillRect(0, 0, raster.getWidth(), raster.getHeight());
+        g.dispose();
+
+        return img;
+    }
+
+    private static ColorSpace createCS(ColorSpaceSelector selector) {
+        return switch (selector) {
+            case GRAY -> ColorSpace.getInstance(ColorSpace.CS_GRAY);
+            case WRAPPED_GRAY -> new TestColorSpace(ColorSpace.getInstance(ColorSpace.CS_GRAY));
+
+            case RGB -> ColorSpace.getInstance(ColorSpace.CS_sRGB);
+            case WRAPPED_RGB -> new TestColorSpace(ColorSpace.getInstance(ColorSpace.CS_sRGB));
+
+            case PYCC -> ColorSpace.getInstance(ColorSpace.CS_PYCC);
+            case WRAPPED_PYCC -> new TestColorSpace(ColorSpace.getInstance(ColorSpace.CS_PYCC));
+        };
+    }
+
+    private static boolean areImagesEqual(BufferedImage destTest, BufferedImage destGold) {
+        for (int x = 0; x < destTest.getWidth(); x++) {
+            for (int y = 0; y < destTest.getHeight(); y++) {
+                int rgb1 = destTest.getRGB(x, y);
+                int rgb2 = destGold.getRGB(x, y);
+                if (rgb1 != rgb2) {
+                    System.err.println("x = " + x + ", y = " + y);
+                    System.err.println("rgb1 = " + Integer.toHexString(rgb1));
+                    System.err.println("rgb2 = " + Integer.toHexString(rgb2));
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    public static void main(String[] args) {
+        BufferedImage srcTest = createTestImage(createCS(ColorSpaceSelector.WRAPPED_GRAY));
+        BufferedImage destTest = createTestImage(createCS(ColorSpaceSelector.WRAPPED_RGB));
+
+        BufferedImage srcGold = createTestImage(createCS(ColorSpaceSelector.GRAY));
+        BufferedImage destGold = createTestImage(createCS(ColorSpaceSelector.RGB));
+
+        ColorSpace mid = createCS(ColorSpaceSelector.PYCC);
+        ColorConvertOp test = new ColorConvertOp(mid, null);
+        test.filter(srcTest, destTest);
+
+        ColorConvertOp gold = new ColorConvertOp(mid, null);
+        gold.filter(srcGold, destGold);
+
+        if (!areImagesEqual(destTest, destGold)) {
+            throw new RuntimeException("ICC test failed");
+        }
+
+        mid = createCS(ColorSpaceSelector.WRAPPED_PYCC);
+        test = new ColorConvertOp(mid, null);
+        test.filter(srcTest, destTest);
+
+        gold = new ColorConvertOp(mid, null);
+        gold.filter(srcGold, destGold);
```

Comment:
Should both filter use the same instance of `ColorConvertOp` where the first one is `gold` and is applied to both `gold-` and `test-` images, then the next instance `ColorConvertOp` is a `test` with the wrapper which is again applied to both images?

Wouldn't it be clearer?

```suggestion
        ColorConvertOp gold = new ColorConvertOp(createCS(ColorSpaceSelector.PYCC), null);
        gold.filter(srcTest, destTest);
        gold.filter(srcGold, destGold);

        if (!areImagesEqual(destTest, destGold)) {
            throw new RuntimeException("ICC test failed");
        }

        ColorConvertOp test = new ColorConvertOp(createCS(ColorSpaceSelector.WRAPPED_PYCC), null);
        test.filter(srcTest, destTest);
        test.filter(srcGold, destGold);
```

Currently, we have `test` and `gold` which use the same `ColorSpace` which could be plain or wrapper for both cases. It looks confusing to me.

What do you think, @mrserb?

Reply:
> Currently, we have test and gold which use the same ColorSpace which could be plain or wrapper for both cases. It looks confusing to me.

That could be fine. We can create a gold image first, then parametrize the test method by different src/mid/dst and compare the result vs gold, so we will cover all possible combinations.

Reply:
> > Currently, we have test and gold which use the same ColorSpace which could be plain or wrapper for both cases. It looks confusing to me.
> 
> That could be fine. We can create a gold image first, then parametrize the test method by different src/mid/dst and compare the result vs gold, so we will cover all possible combinations.

It tried it… with for-loop over `ColorSpaceSelector.PYCC` and `ColorSpaceSelector.WRAPPED_PYCC` and I didn't like the result. Although I kept all the images inside the loop to be safe.

The approach above repeats several lines, however, it's more straightforward and reporting errors is easier: you'll the the line where exception occurred. If it's in the loop, the exception could happen on the same line but with different parameters.